File: code/js/src/http/JsonHome.ts

export class JsonHome {
    constructor(
        api: API,
        resources: Resource[]
    ) {
        this.api = api
        this.resources = resources
    }
}

export interface JsonHome {
    api: API,
    resources: Resource[]
}

export class API {
    constructor(
        title: string,
        links: APILinks
    ) {
        this.title = title
        this.links = links
    }
}

export interface API {
    title: string,
    links: APILinks
}

export class APILinks {
    constructor(
        author: string[],
        describedBy: string
    ) {
        this.author = author
        this.describedBy = describedBy
    }
}

export interface APILinks {
    author: string[],
    describedBy: string
}

export class Resource {
    constructor(
        tag: string,
        hrefTemplate: string,
        hints: Hints
    ) {
        this.tag = tag
        this.hrefTemplate = hrefTemplate
        this.hints = hints
    }
}

export interface Resource {
    tag: string
    hrefTemplate: string,
    hints: Hints
}

export class Hints {
    constructor(
        allow: string[],
        format: string
    ) {
        this.allow = allow
        this.format = format
    }
}

export interface Hints {
    allow: string[],
    format: string
}

File: code/jvm/main/kotlin/com/isel/leic/ps/ionClassCode/infra/JsonHome.kt
package com.isel.leic.ps.ionClassCode.infra

import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity

/**
 * Represents a JsonHome document.
 */
data class JsonHome(
    val api: API,
    val resources: List<Resource>,
)

/**
 * Represents the API section of a JsonHome document.
 */
data class API(
    val title: String = "i-on ClassCode",
    val links: APILinks,
)

/**
 * Represents the links section of a JsonHome document.
 */
data class APILinks(
    val author: List<String> = listOf("A48309@alunos.isel.pt", "A48322@alunos.isel.pt", "A48348@alunos.isel.pt"),
    val describedBy: String = "https://github.com/i-on-project/repohouse/tree/main/docs",
)

/**
 * Represents a resource section of a JsonHome document.
 */
data class Resource(
    val tag: String,
    val hrefTemplate: String,
    val hints: Hints,
)

/**
 * Represents the hints section of a JsonHome document.
 */
data class Hints(
    val allow: List<String>,
    val format: String = "application/json"
)

/**
 * Represents a builder scope for a JsonHome document.
 */
class JsonHomeBuilderScope {

    private val api = API(links = APILinks())
    private val resources = mutableListOf<Resource>()

    fun resource(hrefTemplate: String, tag: String, methods: List<HttpMethod>) {
        resources.add(Resource(hrefTemplate, tag, Hints(allow = methods.map { it.name() })))
    }

    fun build(): JsonHome = JsonHome(api, resources)
}

/**
 * Builds a JsonHome document.
 */
fun jsonHome(block: JsonHomeBuilderScope.() -> Unit): ResponseEntity<JsonHome> {
    val scope = JsonHomeBuilderScope()
    scope.block()
    return ResponseEntity.ok(scope.build())
}
